### PR TITLE
fix: resolve issues with `skuid-cli`

### PR DESCRIPTION
**Pull request by techfg**
> [!IMPORTANT]  
> This PR is a Work-in-progress (WIP) and should **NOT** be merged yet.  Please review details below carefully.

# Jira Issue URL
N/A

# High-Level Description
This PR is a WIP and should not be merged yet.  It contains proposed fixes to many of the issues currently open in the [repo](https://github.com/skuid/skuid-cli) as well as workarounds for issues that aren't directly tied to the CLI itself but exist server side for APIs that the CLI relies on.

Of the 108 issues I've created recently in this repo, this PR currently addresses in full or in-part, 79 of them.  I'm also currently working on addressing approximately 3 more of them (the remaining issues beyond those require Skuid to resolve directly as they either involve documentation updates and/or server-side API fixes).  I hope to have those commits pushed in the next week, time permitting.

I am submitting this PR now vs. waiting until fully complete since there was [recent activity](https://github.com/skuid/skuid-cli/issues/183#issuecomment-2284814332) from Skuid so I'm hopeful that this means other issues will be reviewed as well so want to start the ball rolling on gaining input on the issues themselves and the proposed changes in this PR as I suspect and hope there will be at least some discussion before things can be finalized and merged.

Unfortunately, trying to manage separate PRs for each issue individually would have been nearly impossible given the number of issues present in the cli and the fact that solutions to some issues were dependent on solutions to other issues.  For this reason and while not typical of a PR, I'm submitting a single PR that contains all of them as trying to manage them individually would have been a futile effort.

As referenced throughout the 108 issues, version 0.6.7 of the CLI is not reliable and only under a very specific set of circumstances, does it complete in an expected manner.  In most cases, it either fails entirely or performs unexpectedly.  In short, there are issues in the Docs, issues in the CLI itself and issues in the APIs.  All of these lead to significant risk of unexpected outcomes when deploying to NLX sites, especially given that the CLI is documented as a solution for migrating to `production` environments.

Given all the above, it would be ideal if this PR was reviewed ASAP so solutions for these issues can be officially released and users of the CLI can start to have confidence that management of their NLX sites via the CLI is predictable and reliable.

Regarding the PR itself, a few things:

1.  This is the first time I've used `Go` so while the high number of issues has given me ample opportunity to learn the language, the code should be reviewed closely by someone with at least a mid-level of `Go` experience
2. I am not a Skuid expert and the knowledge I've obtained regarding Skuid is primarily from reviewing the existing CLI code, tinkering via the Web UI admin interface and the CLI itself.  All code changes should be reviewed closely by someone with an expert level of knowledge of Skuid and Skuid Internals.
3. In the absence of a separate PR per issue, I've tried to be fairly verbose in the commit messages to provide context on the changes in the commit so please look there for details on the commit and its included changes.
4. Within the code, where I wasn't 100% sure of something, I included a comment `Skuid Review Required` so when reviewing, beyond general review mentioned in 1 & 2 above, you can specifically look for this comment text to identify where I knew for certain that Skuid input/review was needed.
5. I added lots of tests in nearly every area of code that I touched and, in many cases, more tests than I would typically include in some scenarios, but all are meaningful tests (just a little overkill in some cases).  The reasons for the "thoroughness" in tests is a few-fold:
    1. Version 0.6.7 contained basically no tests, even in areas of the code that contained significant string parsing.  The lack of tests was a primary reason why so many issues exist (as any manual testing that was performed did not detect the issues). This is a fairly significant oversight for an official tool of a commercial product built/supported/maintained by the company of the commercial product and especially one intended to be used for `production purposes`.
    2. In reviewing prior PRs, with some exceptions, there does not appear to have been any type of _code review_ process conducted, at least not one visible via the PR itself.  The only conversation I see around PRs in general is related to "manual testing" and expected/unexpected results.  Given its unclear what type of formal review PRs (either external or internal) undergo, I felt it best to minimize the chances of regressions such as those introduced via [this commit](https://github.com/skuid/skuid-cli/commit/ca75f73a8774d13af9322cf4f8885f797876e195) and [this commit](https://github.com/skuid/skuid-cli/commit/e302ff07c44f509965adbd9adef9cc3874e8242c) which were two of the largest contributors to the issues in the current version of the CLI.
6. This PR is built on top of and includes everything in #137 so #137 could be closed in favor of this one (or both reviewed/merged separately).
7. Breaking Changes:
    1. `Phrasal` support in `--since` has been removed - see comments in [commit message](https://github.com/techfg/skuid-cli/commit/a63d27d60363b74381ce623dabbb7b30da20be86) for details.
    2. Interpretation of `Year/Week/Day` in `--since` parsing is now static values.  It's unclear what the expected behavior for treatment of these options was in 0.6.7 since there were no tests or documentation so this may or may not be a breaking change but mentioning it as a breaking change for clarity.  See comments in [commit message](https://github.com/techfg/skuid-cli/commit/a63d27d60363b74381ce623dabbb7b30da20be86) for details.
8. Non-Breaking Changes:
    1. `--trace`, `--verbose`, `--diagnostic` have been deprecated - see comments in [commit message](https://github.com/techfg/skuid-cli/commit/3bf0a15b57ab405baf20b356cd3f9eea1e22a9ea)
    2. `--no-console-logging` flag added - see [commit](https://github.com/techfg/skuid-cli/commit/816c5d700de05dcdf72d155ab6ba8d119d4b8409)
    3. `--log-level` and `--debug` flag added - see [commit](https://github.com/techfg/skuid-cli/commit/3bf0a15b57ab405baf20b356cd3f9eea1e22a9ea)
    4. `--skip-data-sources` flag added - this is a temporary flag to workaround #150 and can be removed once that issue is resolved.  See [commit](https://github.com/techfg/skuid-cli/commit/47ca287bd428379e143d830916336fba0e2b4e0c).
    5. `--no-clean` flag added - see [commit](https://github.com/techfg/skuid-cli/commit/b599203b4ce241c916703192dda9302d1928fb1b)
    6. All flags now have automatic support for environment variables following the naming convention of `SKUID_FLAG_NAME` (e.g., `SKUID_LOG_DIRECTORY`).  Any previous environment variables that did not follow this pattern have been [deprecated](https://github.com/skuid/skuid-cli/blob/be0028e9b4c4052954b0e235e8ed1a566493195d/pkg/flags/types.go#L26).  See [commit](https://github.com/techfg/skuid-cli/commit/d57819601648c810f5508fe0d30a098d538f92a0) for details.

Looking forward to hearing from Skuid on this PR and the related open issues and helping to bring the CLI to a state of stability and reliability for Skuid customers!

# Changelog:
Individual commits contain references to the issue(s) that it resolves and/or is related to so linking will be done via commit message.  Happy to list them all here once everything is finalized.

In the meantime, here is a spreadsheet that contains a cross-reference to all 108 issues I've created in this repo (+1 that already existed in this repo) along with the commits that target the issue and a "note" column with anything extra that is worth mentioning (e.g., `this is something only skuid can fix`).
[SkuidCLIIssueTracker.xlsx](https://github.com/user-attachments/files/17359451/SkuidCLIIssueTracker.xlsx)

# Additional Information
Below is a summary of high-level changes that were made beyond just the baseline of addressing the core issues/bugs:

1. The previous code had a tremendous amount of string parsing of paths in many different areas of the code - often doing the same "parsing" to condition behavior one way or another. It was a lot of incorrect parsing that led to many of the issues present in 0.6.7. The parsing of paths is now centralized ([here](https://github.com/skuid/skuid-cli/pull/205/files#diff-b84cd94cd3c200927e1546f8478ef1be441127e15e8cb3772f5efc9e2557eaa4R140) and [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-b84cd94cd3c200927e1546f8478ef1be441127e15e8cb3772f5efc9e2557eaa4R160)) and once parsed, strongly typed structures are passed around to avoid the need to reparse and to make conditional execution safe and reliable.
2. All metadata types are now represented as [enums](https://github.com/skuid/skuid-cli/pull/205/files#diff-b84cd94cd3c200927e1546f8478ef1be441127e15e8cb3772f5efc9e2557eaa4R70) so there is no longer a need for re-parsing to detect what metadata type something is - another area that led to many issues in the existing code.
3. Dependency injection was introduced in several areas (e.g. [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-d90a10e0e4ef415fb401a1192be528cc61c01c329bf9d9133934426ea07d7d31R8)) that required significant refactoring to resolve existing issues.  This improves testability (see below) and sets the table for further use of DI for things like HttpClient, etc. to enable improving test coverage throughout the CLI.
4. Mocking was introduced via the [mockery](https://github.com/skuid/skuid-cli/pull/205/files#diff-531f7a01bb4f278f511fe2ada0896cfbbcb0771b7e53093d3a90c425882e4994R1) package to support testing of DI related items.
5. Code testability improvements - Were new code was added and/or significant changes required to address issues, code was refactored to improve testability.  A large majority of the code in 0.6.7 is not testable short of writing full integration tests because: 1) the logic in the code isn't separated into testable units; 2) there is no use of dependency injection so anything that exercises things like HttpClient must use a real HttpClient (vs. a mock).  Additionally, the integration tests that do exist fall short of doing much in the way of being "meaningful" (e.g., they [deploy](https://github.com/skuid/skuid-cli/blob/master/pkg/deploy_test.go#L13) to a server and simply check that no error was returned vs ensuring that the server contains the expected result which theoretically may be ok if there were other unit tests that existed to ensure proper behavior).  The lack of tests and lack of testability of the code is something that likely (hopefully) would have been caught in a _code review_ process (see mention of this above in `Regarding the PR itself, a few things` Item 5.2.  Some examples of these changes are:
    1. The [code](https://github.com/skuid/skuid-cli/blob/master/cmd/retrieve.go#L131) to "parse" the `--since` flag is buried inside of a much larger function that orchestrates the entire retrieval process.  There is no way, short of full integration tests of which there would have to be many to test all the scenarios of parsing `since` to test that parsing `since` behaves as expected.  This code was refactored [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-038d3efb4a63a69256f41a35ab55a1cf5b4e7bf66d9e3a9cbc172889e3a281acR21) and [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-b85cb8dd9eecd095561611310332de19cd87bb74b60f8083131caec5e05d0b0dR44) and tests written against behaviors [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-1b434e4a3265e89031c9633143a5101a69c70619ad92fdc155418983e88e6ec4R13) and [here](https://github.com/skuid/skuid-cli/pull/205/files#diff-a5e9a76637698de0da6211355a3a524a5399e131c15664fe3060e9a816982a71R66).  
    3. The [code](https://github.com/skuid/skuid-cli/blob/master/pkg/zip.go#L20) to archive files contained essentially no tests despite a significant amount of conditional logic, filtering, etc.  While the code was testable, writing tests was not straightforward since the only way to write tests was to utilize the physical file system and any test setup would be dependent on fixtures that existed in the repo.  The [refactored approach](https://github.com/skuid/skuid-cli/blob/be0028e9b4c4052954b0e235e8ed1a566493195d/pkg/zip.go#L89) utilizes the `fs` package allowing for `MapFS` to be used, dependency injection (e.g., `fs.FS`, [FileUtil](https://github.com/skuid/skuid-cli/blob/be0028e9b4c4052954b0e235e8ed1a566493195d/pkg/util/files.go#L35)) and separated `filters` (each themselves individually testable  e.g., [MetadataArchiveFilter](https://github.com/skuid/skuid-cli/blob/be0028e9b4c4052954b0e235e8ed1a566493195d/pkg/zip.go#L64)).  In short, test coverage went from [near 0%](https://github.com/skuid/skuid-cli/blob/master/pkg/zip_test.go#L13) to [100%](https://github.com/skuid/skuid-cli/blob/be0028e9b4c4052954b0e235e8ed1a566493195d/pkg/zip_test.go) for everything related to `Archive`.

### CODE REVIEW

**Comment by techfg** [comment]
Updated description with additional information and updated tracking spreadsheet.  Will continue to update both as I think of more relevant info and as further commits are added to PR.

**Comment by techfg** [comment]
Updated description as follows:

1. Found more bugs, created new issues (#206, #207, #208), updated spreadsheet & PR description accordingly
2. Addressed issues (#157, #207, #208), pushed commits, updated spreadsheet
3. attached updated spreadsheet

**Comment by techfg** [comment]
1. Filed a couple more bugs, creating issues #213 & #214
2. Implemented the following:
    1. #169 - Added ci workflow to test macOS, ubuntu & windows, note that this workflow when run against current version 0.6.7 fails on windows due to #214
    2. #213 - Temporary workaround until expected behavior is understood
    3. #214 - `FromWindowsPath` function and its tests were removed in 8478de11f0eb726f87edafceb95030af3a630059 because it's not even needed
3. Updated PR description reflecting above including attaching updated spreadsheet

All other information from previous comment still applies.

**Comment by techfg** [comment]
1. Found a couple more bugs, created issues #215, #216
2. Addressed #215
3. Updated PR description reflecting above including attaching updated spreadsheet

All other information from [this comment](https://github.com/skuid/skuid-cli/pull/205#issuecomment-2352388060) still applies.

**Comment by techfg** [comment]
**TL;DR**
1. Created 19 new issues (14 bugs, 1 feature, 4 chores) - #217, #218, #219, #220, #221, #222, #223, #224, #225, #226, #227, #228, #229, #230, #231, #232, #233, #234, #235
2. Addressed 16 of the 19 issues - #217, #218, #219, #220, #221, #222, #223, #224, #225, #226, #227, #228, #229, #230, #231, #233
3. Updated PR description reflecting above including attaching updated spreadsheet

Up until this batch of issues/commits, I have avoided any type of significant refactoring of the existing code base (except for where it was necessary to resolve an issue(s)) due to the high risk of regression that may be introduced given the lack of unit tests and the high complexity of the code structure as written (e.g., 200+ line methods with many different things all being performed inline).  

However, given all the issues still present even with all the fixes/improvements in this PR, having meaningful log messages was critical to be able to identify where things are not working as expected. As an example, if v0.6.7 had anything resembling adequate logging such as what now exists in this PR, identifying root cause of issues that have been found would have been much simpler.

To that end, to address #220, refactoring the code in the remaining areas that hadn't been refactored yet was necessary in order to not further complicate an already unnecessarily complex/difficult to follow code base.  Along that line, #230 was created so that #220 & #230 could be addressed in tandem as they are essential for each other to produce the originally desired outcome of #220.  

This [commit](https://github.com/techfg/skuid-cli/commit/af20e1e74ff14ee0bfeab32a2c95f060670fe3a0) covers a large amount of ground issue wise as a single commit as it would have been nearly impossible to split up into smaller commits since logging is foundational to everything.  It contains the refactor itself along with addressing issues related to logging and many bugs found during the refactor while significantly improving readability of the code, thus lowering the overall cost of maintainability.

Below is a summary of what has been accomplished in the 16 issues addressed since my last PR comment above:

### Logging

> [!NOTE]
> Due to all the issues with v0.6.7 and all the open issues in this repo, logging is very extensive as it's required for two primary reasons:
> 
> 1. PR (#205) is massive, and it will help with the review process
> 2. There are still many issues that remain (in the CLI itself and in the server-side APIs) so having the extensive logging is the only way to even try to understand what went wrong during command execution
> 
> Once the outstanding server API issues are resolved and the CLI stable, the amount of logging can be reduced to a more "reasonable" level

- all messages follow same format
- when file logging enabled, all messages are logged to file (even panics) except for in the situation where an error/panic occurs prior to root command's `PersistentPreRunE` being invokved
- all durations & times are formatted consistently and all fields (--diag) are quoted to ensure consistency in their output
- all higher level activities and several lower level activities are traced/timed
- Fields (when `--diag` is passed) contain several key/value pairs consistently for every message:
  - `logger` - the name of the function
  - `error` - if an error was present when message was logged
  - `command` - if the function is the "entry" for a command
  - if the function logging the message is "under tracking" control:
    - `start` - the time the function started tracking
    - `finish` - the time the function finished tracking
    - `duration` - the duration of the function (in nanoseconds)
    - `success` - false if the function returned error
- use of color is intentional and carries a meaning (see [message.go](https://github.com/techfg/skuid-cli/blob/tfgbeta/pkg/logging/message.go#L23))
- eliminated all occurrences of misuse of "WithFields"
- use of debug & trace messages is clearly defined and message/message format/fields/etc. are identical between them - debug is used for higher-level activities, trace is used for individual plan/entity/file/etc. activities
- Anything that "may" contain sensitive information requires the "--diag" flag to be passed (code in this commit should not log anything sensitive unless APIs return sensitive information in error messages which hopefully they do not)
- Deploy & Retrieve commands log all names of entities deployed/retrieved in aggregate and per plan when `--log-level` is `debug` (or greater) - note that due to #211, the entity name results from `execute deploy plan` are not logged because the results returned are incomplete, inaccurate, and/or not existent.

### Log Files
Log files can now be evaluated using various tools due to having a consistent format.  For example:

- the following command can be executed to see the duration of every tracked function, returning the duration (nanoseconds), logger/function name and message:
    ```bash
    grep 'duration="[^"]*"' skuid-cli-2024-10-03T031319-0700-2495227290.log | sed -E 's/.*msg=("[^"]*").*duration="([^"]*)".*logger="([^"]*)".*/\2\t | \3\t | \1/' | sort -h | column -s $'\t' -t
    ```
    ```text
    # Retrieve Example
    ...
    834986070    | pkg.Authorize                                                           | "SUCCESS Requesting authentication in 834.98607ms"
    990743036    | pkg.ExecuteRetrieval                                                    | "SUCCESS Executing retrieval plan(s) in 990.743036ms"
    1201127548   | pkg.Retrieve                                                            | "SUCCESS Retrieving site https://<redacted>.skuidsite.com to directory `/sites/mysite/skuid-objects` in 1.201127548s"
    2036255161   | cmd.retrieve                                                            | "SUCCESS Executing command `retrieve` entities from site https://<redacted>.skuidsite.com to directory `/sites/mysite/skuid-objects` in 2.036255161s"
    2036255161   | cmd.retrieve                                                            | "✓ Retrieved site https://<redacted>.skuidsite.com to `/sites/mysite/skuid-objects`"

    # Deploy Example
    ...
    1986807659   | pkg.ExecuteDeployPlan.pkg.executeDeployPlan                             | "SUCCESS Executing deployment for plan `Skuid NLX` in 1.986807659s"
    3039649260   | pkg.ExecuteDeployPlan                                                   | "SUCCESS Executing deployment plan(s) in 3.03964926s"
    3362697168   | pkg.Deploy                                                              | "SUCCESS Deploying site https://<redacted>.skuidsite.com from directory `/sites/mysite/skuid-objects` in 3.362697168s"
    3940579660   | cmd.deploy                                                              | "SUCCESS Executing command `deploy` entities to site https://<redacted>.skuidsite.com from directory `/sites/mysite/skuid-objects` in 3.94057966s"
    3940579660   | cmd.deploy                                                              | "✓ Deployed site https://<redacted>.skuidsite.com from `/sites/mysite/skuid-objects`"

    ```
- the following command can be executed to see the aggregate and per plan entities deployed/retrieved (example is from `deploy`):
    ```bash
    grep 'entities="[^"]*"' skuid-cli-2024-10-03T221256-0700-3834979381.log | sed -E 's/.*entities="([^"]*)".*entitiesFrom="([^"]*)".*/\2\t | \1/' | column -s $'\t' -t
    ```
    ```text
    # Retrieve Example
    Execute retrieval plan(s) [`Skuid Cloud Data Service`, `Skuid NLX`]          | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute retrieval plan `Skuid NLX` request                                   | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute retrieval plan `Skuid Cloud Data Service` request                    | [`apps/SkuidSample`, `apps/sampleapp`, `sitepermissionsets/Admin`, `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute retrieval plan `Skuid NLX` response                                  | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute retrieval plan `Skuid Cloud Data Service` response                   | [`sitepermissionsets/Admin`]
    Execute retrieval plan(s) [`Skuid Cloud Data Service`, `Skuid NLX`] result   | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]


    # Deploy Example
    --entities flag                                                        | []
    Get deployment plan(s) request                                         | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute deployment plan(s) [`Skuid Cloud Data Service`, `Skuid NLX`]   | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute deployment plan `Skuid NLX` request                            | [`apps/SkuidSample`, `apps/sampleapp`, `datasources/SkuidSample_NW_OData`, `designsystems/SkuidSample NW`, ..., `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Execute deployment plan `Skuid Cloud Data Service` request             | [`apps/SkuidSample`, `apps/sampleapp`, `sitepermissionsets/Admin`, `sitepermissionsets/Public`, `sitepermissionsets/Standard`]
    Update Permission Sets request                                         | [`test permission set`]
    --entities flag                                                        | []
    --entities flag                                                        | []
    ```

## Coding Structure/Style
- refactored functions to have a single purpose or orchestrate smaller composable units via function calls to improve overall clarity of purpose and decrease cost of maintenance (and risks)
- No naked returns
- Named returns when needed to clarify meaning of a return value and/or required in a defer function

## General
- general performance improvements (e.g., eliminated copying entire data structures just to change a single property to pass to another function, migrated to github.com/goccy/go-json for all json operations)
- eliminated several scenarios where very fragile conditional logic was being applied to make decisions, changing the approach to being explicit/intentional and failing when something unexpected was encountered
- retrieve & deploy commands now use the same data structure for NlxPlan for consistency and also reducing risk of incorrect usage of plans as the order of evaluation/execution of plans is critical and should be very intentional

Resolves #59
Resolves #165
Resolves #199
Resolves #217
Resolves #218
Resolves #219
Resolves #220
Resolves #221
Resolves #222
Resolves #223
Resolves #224
Resolves #225
Resolves #226
Resolves #227
Resolves #228
Resolves #229
Resolves #230
Resolves #231
Resolves #233
